x/reward/types/action_builder.go

```diff
@@ -0,0 +1,240 @@
+package types
+
+import (
+	"fmt"
+
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/cosmos/cosmos-sdk/x/gov/types"
+)
+
+var (
+	_ ActionBuilder = &TransferActionBuilder{}
+	_ ActionBuilder = &DelegateActionBuilder{}
+	_ ActionBuilder = &VoteActionBuilder{}
+)
+
+type ActionBuilder interface {
+	GetEventCriteria() *EventCriteria
+	AddEvent(eventType string, attributes *map[string][]byte) error
+	CanBuild() bool
+	BuildAction() (EvaluationResult, error)
+	Reset()
+}
+
+type TransferActionBuilder struct {
+	Sender    sdk.AccAddress
+	Action    string
+	Recipient sdk.AccAddress
+}
+
+func (b *TransferActionBuilder) GetEventCriteria() *EventCriteria {
+	return NewEventCriteria([]ABCIEvent{
+		{
+			Type: sdk.EventTypeMessage,
+			Attributes: map[string][]byte{
+				sdk.AttributeKeyAction: []byte(sdk.MsgTypeURL(&banktypes.MsgSend{})),
+			},
+		},
+		{
+			Type:       banktypes.EventTypeTransfer,
+			Attributes: map[string][]byte{},
+		},
+	})
+}
+
+func (b *TransferActionBuilder) AddEvent(eventType string, attributes *map[string][]byte) error {
+	switch eventType {
+	case sdk.EventTypeMessage:
+		// Update the result with the senders address
+		address := (*attributes)[banktypes.AttributeKeySender]
+		action := (*attributes)[sdk.AttributeKeyAction]
+		if address != nil {
+			address, err := sdk.AccAddressFromBech32(string(address))
+			if err != nil {
+				return err
+			}
+			b.Sender = address
+		}
+
+		if action != nil {
+			b.Action = string(action)
+		}
+	case banktypes.EventTypeTransfer:
+		// Update the result with the senders address
+		address := (*attributes)[banktypes.AttributeKeySender]
+		addressSenderAddr, err := sdk.AccAddressFromBech32(string(address))
+		if err != nil {
+			return err
+		}
+		b.Sender = addressSenderAddr
+
+		addressRecipient := (*attributes)[banktypes.AttributeKeyRecipient]
+		addressRecipientAddr, errFromParsingRecipientAddress := sdk.AccAddressFromBech32(string(addressRecipient))
+		if errFromParsingRecipientAddress != nil {
+			return errFromParsingRecipientAddress
+		}
+		b.Recipient = addressRecipientAddr
+	}
+	return nil
+}
+
+func (b *TransferActionBuilder) CanBuild() bool {
+	return !b.Sender.Empty() && !(len(b.Action) == 0) && !b.Recipient.Empty()
+}
+
+func (b *TransferActionBuilder) BuildAction() (EvaluationResult, error) {
+	if !b.CanBuild() {
+		return EvaluationResult{}, nil
+	}
```

Comment:
The other action builders return an error in this case. Should this one too or is there something special for this one? If special, a comment would help.

Reply:
should return error..changed